certbot/plugins/dns_common.py

```diff
@@ -0,0 +1,318 @@
+        """
+
+        def __validator(filename):
+            if required_variables:
+                CredentialsConfiguration(filename, self.dest).require(required_variables)
+
+        self._configure_file(key, label, __validator)
+
+        credentials_configuration = CredentialsConfiguration(self.conf(key), self.dest)
+        if required_variables:
+            credentials_configuration.require(required_variables)
+
+        return credentials_configuration
+
+    @staticmethod
+    def _prompt_for_data(label):
+        """
+        Prompt the user for a piece of information.
+
+        :param str label: The user-friendly label for this piece of information.
+        :returns: The user's response (guaranteed non-empty).
+        :rtype: str
+        """
+
+        def __validator(i):
+            if not i:
+                raise errors.PluginError('Please enter your {0}.'.format(label))
+
+        code, response = ops.validated_input(
+            __validator,
+            'Input your {0}'.format(label),
+            force_interactive=True)
+
+        if code == display_util.OK:
+            return response
+        else:
+            raise errors.PluginError('{0} required to proceed.'.format(label))
+
+    @staticmethod
+    def _prompt_for_file(label, validator=None):
+        """
+        Prompt the user for a path.
+
+        :param str label: The user-friendly label for the file.
+        :param callable validator: A method which will be called to validate the supplied input
+            after it has been validated to be a non-empty path to an existing file. Should throw a
+            `~certbot.errors.PluginError` to indicate any issue.
+        :returns: The user's response (guaranteed to exist).
+        :rtype: str
+        """
+
+        def __validator(filename):
+            if not filename:
+                raise errors.PluginError('Please enter a valid path to your {0}.'.format(label))
+
+            filename = os.path.expanduser(filename)
+
+            validate_file(filename)
+
+            if validator:
+                validator(filename)
+
+        code, response = ops.validated_directory(
+            __validator,
+            'Input the path to your {0}'.format(label),
+            force_interactive=True)
+
+        if code == display_util.OK:
+            return response
+        else:
+            raise errors.PluginError('{0} required to proceed.'.format(label))
+
+
+class CredentialsConfiguration(object):
+    """Represents a user-supplied filed which stores API credentials."""
+
+    def __init__(self, filename, mapper=lambda x: x):
+        """
+        :param str filename: A path to the configuration file.
+        :param callable mapper: A transformation to apply to configuration key names
+        :raises PluginError: If the file does not exist.
+        """
+        validate_file_permissions(filename)
+
+        self.confobj = configobj.ConfigObj(filename)
```

Comment:
If the file isn't a valid ini file, Certbot crashes with a traceback from `configobj`. Perhaps `configobj` raising an error should be documented in the docstring here and a validator that creates this object can catch (and log) this exception and explain the problem to the user?